### PR TITLE
test: add max block size test using access lists

### PR DESCRIPTION
**Pull request by nerolation**
Add test_block_full_access_list_and_data to test maximum block size by splitting gas between access lists (60%) and calldata (40%). Uses 1 address with many storage keys to maximize block size.

## 🔗 Related Issues or PRs

N/A.

## ✅ Checklist

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx --with=tox-uv tox -e lint,typecheck,spellcheck,markdownlint
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [ ] All: Considered adding an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [ ] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-spec-tests/blob/main/docs/writing_tests/post_mortems.md) to add an entry the list.
- [ ] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-spec-tests/blob/main/tests/static) have been assigned `@ported_from` marker.

### CODE REVIEW

**Comment by LouisTsai-Csie** [comment]
Nice work! I am just curious why the ratio is `access lists`: `calldata` = 60%: 40% here?

**Comment by nerolation** [comment]
This is because under EIP-7623, you only hit the floor price for calldata if you use <60% of the available gas for EVM execution.
2930 access lists are interesting: One the one hand, they behave like calldata (add a lot of data to the block) but on the other hand we price them for their storage access. I have an EIP to change that: https://eips.ethereum.org/EIPS/eip-7981

Now, because we treat ALs as "storage access", they also contribute to the EVM part for calldata pricing. Thus, the max possible block size (snappy compresses) is the one I got in the test case.
Actually, we could optimize for a large size even further by using 29% of the calldata gas available for zero-bytes and the rest for non-zero bytes. This is where snappy compression does its worse job.

Added it in the last commit here.

**Comment by nerolation** [comment]
Adding this here, explaining the 29/71 ratio 
70671ad2-cec9-44d6-8719-6bc1dc7715c0

File: tests/benchmark/test_worst_blocks.py [hunk 1]
```diff
@@ -222,3 +230,94 @@ def test_block_full_data(
...
+@pytest.mark.valid_from("Prague")
+def test_block_full_access_list_and_data(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    intrinsic_cost: int,
+    total_cost_standard_per_token: int,
+    fork: Fork,
+):
+    """Test a block with access lists (60% gas) and calldata (40% gas) using random mixed bytes."""
+    import random
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:244]: Normally we try to keep the imports at the top of the file.

File: tests/benchmark/test_worst_blocks.py [hunk 2]
```diff
@@ -222,3 +230,94 @@ def test_block_full_data(
...
+    
+    attack_gas_limit = Environment().gas_limit
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:246]: On latest main we revised this approach and now this is the expected way of getting the maximum benchmarking gas:
```python
@pytest.mark.valid_from("Prague")
def test_block_full_access_list_and_data(
    state_test: StateTestFiller,
    pre: Alloc,
    intrinsic_cost: int,
    total_cost_standard_per_token: int,
    fork: Fork,
    gas_benchmark_value: int,  # New parameter, passed by the benchmark tests filler
):
    attack_gas_limit = gas_benchmark_value
```

File: tests/benchmark/test_worst_blocks.py [hunk 3]
```diff
@@ -222,3 +230,94 @@ def test_block_full_data(
...
+    num_non_zero_bytes = tokens_for_non_zero_bytes // 4  # 4 tokens = 1 non-zero byte
+    
+    # Create calldata with mixed bytes
+    calldata = bytearray()
+    
+    # Add zero bytes
+    calldata.extend(b"\x00" * num_zero_bytes)
+    
+    # Add non-zero bytes (random values from 0x01 to 0xff)
+    random.seed(42)  # For reproducibility
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:301]: We normally use a random object that contains a separate state not shared with other tests potentially running in parallel:
```suggestion
    rng = random.Random(42)  # For reproducibility
```
And then this can be used as normal in place of `random.randint` and `random.shuffle`:
```python
calldata.append(rng.randint(1, 255))
```

File: tests/benchmark/test_worst_blocks.py [hunk 4]
```diff
@@ -222,3 +230,94 @@ def test_block_full_data(
...
+    tx = Transaction(
+        to=pre.fund_eoa(),
+        data=calldata,
+        gas_limit=attack_gas_limit,
+        sender=pre.fund_eoa(),
+        access_list=access_list,
+    )
+
+    state_test(
+        env=Environment(),
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:319]:
```suggestion
```
We can remove this line since the environment for all benchmark tests is now prepared by the test filler.

File: tests/benchmark/test_worst_blocks.py [hunk 5]
```diff
@@ -222,3 +230,94 @@ def test_block_full_data(
...
+    
+    # Shuffle the bytes to mix zero and non-zero bytes
+    calldata_list = list(calldata)
+    random.shuffle(calldata_list)
+    calldata = bytes(calldata_list)
+
+    tx = Transaction(
+        to=pre.fund_eoa(),
+        data=calldata,
+        gas_limit=attack_gas_limit,
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:313]: I think this case in particular is affected by Osaka's EIP-7825, and we might need to update it in a future PR to make it multi-tx.

File: tests/benchmark/test_worst_blocks.py [hunk 6]
```diff
@@ -222,3 +230,94 @@ def test_block_full_data(
...
+    for _ in range(num_non_zero_bytes):
+        calldata.append(random.randint(1, 255))
+    
+    # Shuffle the bytes to mix zero and non-zero bytes
+    calldata_list = list(calldata)
+    random.shuffle(calldata_list)
+    calldata = bytes(calldata_list)
+
+    tx = Transaction(
+        to=pre.fund_eoa(),
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:311]:
```suggestion
        to=pre.fund_eoa(amount=0),
```
Just a small trick to make this compatible with `execute` command, to not have it fund an account that's will not be used to send transactions.

File: tests/benchmark/test_worst_blocks.py (hunk 5 above)
Comment (LouisTsai-Csie) [reply tests/benchmark/test_worst_blocks.py:318]: This feature is scheduled in PR https://github.com/ethereum/execution-spec-tests/pull/1945

File: tests/benchmark/test_worst_blocks.py [hunk 7]
```diff
@@ -219,3 +229,92 @@ def test_block_full_data(
...
+        data=shuffled_calldata,
+        gas_limit=attack_gas_limit,
+        sender=pre.fund_eoa(),
+        access_list=access_list,
+    )
+
+    state_test(
+        pre=pre,
+        post={},
+        tx=tx,
```
Comment (marioevz) [inline tests/benchmark/test_worst_blocks.py:319]:
```suggestion
        tx=tx,
        expected_benchmark_gas_used=fork.transaction_intrinsic_cost_calculator()(
            calldata=shuffled_calldata,
            access_list=access_list,
        ),
```
One last fix to make tests pass post-refactor! :)
